File: TheTracker/ui/meta.js
// Meta pages: which heroes are strong right now, and which way they moved.
//
// The shape of these pages is borrowed from dota2protracker and Statlocker —
// heroes grouped and ranked, a trend you can read at a glance, items that
// are rising. None of the numbers are theirs: everything here is computed
// in meta.rs from OpenDota and the community Deadlock API. See that file
// for why the line is drawn where it is.

const META = {
  dota: { data: null, loadedAt: 0, loading: false, error: null, role: "all", sort: "winRate" },
  dl: { data: null, loadedAt: 0, loading: false, error: null, tab: "heroes", sort: "winRate" },
};

// The backend caches the expensive aggregate calls for thirty minutes. Keep
// the browser cache on the same schedule; before this, opening Meta once
// meant it silently stayed stale for the rest of the app session.
const META_CACHE_MS = 30 * 60 * 1000;

function metaUpdatedAt(timestamp) {
  if (!timestamp) return "Updated now";
  const minutes = Math.max(0, Math.floor((Date.now() - timestamp) / 60000));
  return minutes ? `Updated ${minutes}m ago` : "Updated just now";
}

const META_SORTS = [
  ["winRate", "Win rate"],
  ["pickRate", "Picked"],
  ["trend", "Trending"],
];

function metaPct(n, digits = 1) {
  return `${n.toFixed(digits)}%`;
}

function metaCount(n) {
  if (n >= 1000000) return `${(n / 1000000).toFixed(1)}M`;
  if (n >= 1000) return `${(n / 1000).toFixed(n >= 10000 ? 0 : 1)}K`;
  return String(n);
}

/// A signed win-rate move, shown only when it is big enough to mean
/// anything. Under a third of a point is noise in a sample this size, and
/// an arrow on every row would make the ones that matter invisible.
function metaTrend(delta) {
  if (delta === null || delta === undefined || Math.abs(delta) < 0.35) {
    return `<span class="meta-trend flat">—</span>`;
  }
  const up = delta > 0;
  return `<span class="meta-trend ${up ? "up" : "down"}">${up ? "▲" : "▼"} ${Math.abs(delta).toFixed(1)}</span>`;
}

/// Win rate drives the bar, but a bar from 0–100% would put every hero in
/// the middle: real win rates live between about 42% and 58%. The scale is
/// clamped to that band so the differences that matter are visible.
function metaBar(winRate) {
  const pos = Math.max(0, Math.min(1, (winRate - 42) / 16)) * 100;
  const tone = winRate >= 52 ? "good" : winRate >= 48 ? "even" : "bad";
  return `<div class="meta-bar"><div class="meta-bar-fill ${tone}" style="width:${pos}%"></div></div>`;
}

// ---------- Dota ----------

async function loadDotaMeta(force = false) {
  if (META.dota.loading) return;
  if (META.dota.data && !force && Date.now() - META.dota.loadedAt < META_CACHE_MS) return;
  META.dota.loading = true;
  META.dota.error = null;
  renderDotaMeta();
  try {
    META.dota.data = await invoke("dota_meta");
    META.dota.loadedAt = Date.now();
  } catch (e) {
    META.dota.error = String(e);
  }
  META.dota.loading = false;
  renderDotaMeta();
}

/// Descending on every column here — win rate, pick rate and trend all read
/// "most first". A missing value sorts to the bottom rather than the top.
function metaSorted(list, key) {
  return [...list].sort((a, b) => (b[key] ?? 0) - (a[key] ?? 0));
}

function renderDotaMeta() {
  const root = document.getElementById("tab-dotameta");
  if (!root) return;

  if (META.dota.error) {
    root.innerHTML = `
      <section class="home-section" style="padding-top:0">
        <div class="home-head"><h2 class="home-title">Couldn't load the meta</h2></div>
        <p class="hint">${escapeHtml(META.dota.error)}</p>
        <button class="btn btn-secondary" style="margin-top:12px" data-meta-retry type="button">Try again</button>
      </section>`;
    root.querySelector("[data-meta-retry]").addEventListener("click", () => loadDotaMeta(true));
    return;
  }

  const d = META.dota.data;
  if (!d) {
    root.innerHTML = `<div class="empty-state">${META.dota.loading ? "Reading the meta…" : ""}</div>`;
    return;
  }

  const role = META.dota.role;
  const pool = role === "all" ? d.heroes : d.heroes.filter((h) => h.roles.includes(role));
  const rows = metaSorted(pool, META.dota.sort).slice(0, 40);

  // Movers are taken from heroes with a real sample, so a hero picked twice
  // last week cannot top the list on a fluke.
  const solid = d.heroes.filter((h) => h.pickRate >= 1.5);
  const rising = metaSorted(solid, "trend").slice(0, 5);
  const falling = [...solid].sort((a, b) => a.trend - b.trend).slice(0, 5);
  const contested = [...d.heroes].sort((a, b) => b.pickRate - a.pickRate).slice(0, 5);

  const moverRow = (h) => `
    <div class="meta-mover static">
      <img class="hero-portrait" src="${DOTA_HERO_CDN}${escapeHtml(h.slug)}.png" alt="" loading="lazy" decoding="async" />
      <span class="meta-mover-name">${escapeHtml(h.name)}</span>
      <span class="meta-mover-wr">${metaPct(h.winRate)}</span>
      ${metaTrend(h.trend)}
    </div>`;

  root.innerHTML = `
    <section class="home-section" style="padding-top:0">
      <div class="home-head">
        <h2 class="home-title">The meta right now</h2>
        <div class="home-meta">public matches · ${metaUpdatedAt(META.dota.loadedAt)}</div>
        <button class="link-btn" data-meta-retry type="button">${META.dota.loading ? "Refreshing…" : "Refresh"}</button>
      </div>
      ${railHtml([
        { label: "Sample", value: metaCount(d.matches), sub: "matches analysed" },
        { label: "Heroes", value: String(d.heroes.length), sub: "with a pick sample" },
        {
          label: "Most contested",
          value: escapeHtml(contested[0] ? contested[0].name : "—"),
          sub: contested[0] ? `${metaPct(contested[0].pickRate, 0)} of matches` : "",
        },
      ])}
      <p class="hint" style="margin-top:12px;max-width:74ch">
        Picks and wins across public matches, from OpenDota. Trend compares
        the recent half of the sample window against the earlier half, so a
        hero has to move for several days before it shows.
      </p>
    </section>

    <section class="home-section">
      <div class="home-head"><h2 class="home-title">On the way up</h2><div class="home-meta">win rate, last few days</div></div>
      <div class="meta-movers">${rising.map(moverRow).join("")}</div>
    </section>

    <section class="home-section">
      <div class="home-head"><h2 class="home-title">On the way down</h2><div class="home-meta">win rate, last few days</div></div>
      <div class="meta-movers">${falling.map(moverRow).join("")}</div>
    </section>

    <section class="home-section">
      <div class="home-head">
        <h2 class="home-title">Every hero</h2>
        <div class="home-meta">${pool.length} shown</div>
      </div>

      <div class="chip-row" style="margin-bottom:10px">
        <button class="chip ${role === "all" ? "selected" : ""}" data-meta-role="all" type="button">All roles</button>
        ${d.roles
          .map(
            (r) =>
              `<button class="chip ${role === r ? "selected" : ""}" data-meta-role="${escapeHtml(r)}" type="button">${escapeHtml(r)}</button>`
          )
          .join("")}
      </div>
      <div class="chip-row" style="margin-bottom:14px">
        ${META_SORTS.map(
          ([k, label]) =>
            `<button class="chip ${META.dota.sort === k ? "selected" : ""}" data-meta-sort="${k}" type="button">${label}</button>`
        ).join("")}
      </div>

      <div class="meta-head">
        <span></span><span>Hero</span><span class="num">Win rate</span><span></span>
        <span class="num">Picked</span><span class="num">Divine+</span><span class="num">Pro</span><span class="num">Trend</span>
      </div>
      ${rows
        .map(
          (h) => `
        <div class="meta-row static">
          <img class="hero-portrait" src="${DOTA_HERO_CDN}${escapeHtml(h.slug)}.png" alt="" loading="lazy" decoding="async" />
          <div class="meta-name">
            ${escapeHtml(h.name)}
            <span class="meta-roles">${h.roles.slice(0, 2).map(escapeHtml).join(" · ")}</span>
          </div>
          <span class="num meta-wr ${h.winRate >= 50 ? "good" : "bad"}">${metaPct(h.winRate)}</span>
          ${metaBar(h.winRate)}
          <span class="num">${metaPct(h.pickRate, 1)}</span>
          <span class="num">${h.highWinRate === null || h.highWinRate === undefined ? "—" : metaPct(h.highWinRate)}</span>
          <span class="num">${h.proPicks ? `${h.proPicks}p / ${h.proBans}b` : "—"}</span>
          <span class="num">${metaTrend(h.trend)}</span>
        </div>`
        )
        .join("")}
      <p class="hint" style="margin-top:12px;max-width:74ch">
        Roles are Valve's own tags, not positions one through five — OpenDota
        does not publish position data, and guessing it would be worse than
        saying so. <b>Divine+</b> is the highest rank bracket with a sample.
      </p>
    </section>`;

  root.querySelectorAll("[data-meta-role]").forEach((el) =>
    el.addEventListener("click", () => {
      META.dota.role = el.dataset.metaRole;
      renderDotaMeta();
    })
  );
  root.querySelectorAll("[data-meta-sort]").forEach((el) =>
    el.addEventListener("click", () => {
      META.dota.sort = el.dataset.metaSort;
      renderDotaMeta();
    })
  );
  root.querySelectorAll("[data-meta-retry]").forEach((el) =>
    el.addEventListener("click", () => loadDotaMeta(true))
  );
}

// ---------- Deadlock ----------

async function loadDeadlockMeta(force = false) {
  if (META.dl.loading) return;
  if (META.dl.data && !force && Date.now() - META.dl.loadedAt < META_CACHE_MS) return;
  META.dl.loading = true;
  META.dl.error = null;
  renderDeadlockMeta();
  try {
    META.dl.data = await invoke("deadlock_meta");
    META.dl.loadedAt = Date.now();
  } catch (e) {
    META.dl.error = String(e);
  }
  META.dl.loading = false;
  renderDeadlockMeta();
}

function renderDeadlockMeta() {
  const root = document.getElementById("tab-dlmeta");
  if (!root) return;

  if (META.dl.error) {
    root.innerHTML = `
      <section class="home-section" style="padding-top:0">
        <div class="home-head"><h2 class="home-title">Couldn't load the meta</h2></div>
        <p class="hint">${escapeHtml(META.dl.error)}</p>
        <button class="btn btn-secondary" style="margin-top:12px" data-meta-retry type="button">Try again</button>
      </section>`;
    root.querySelector("[data-meta-retry]").addEventListener("click", () => loadDeadlockMeta(true));
    return;
  }

  const d = META.dl.data;
  if (!d) {
    root.innerHTML = `<div class="empty-state">${META.dl.loading ? "Reading the meta…" : ""}</div>`;
    return;
  }

  const heroes = [...d.heroes].sort((a, b) =>
    META.dl.sort === "pickRate" ? b.pickRate - a.pickRate : b.winRate - a.winRate
  );
  const top = heroes[0];
  const contested = [...d.heroes].sort((a, b) => b.pickRate - a.pickRate)[0];

  // Ranked by win rate, but only once an item is common enough to mean
  // something — a niche pick in a handful of games would otherwise top the
  // table. Share is measured against the most-bought item, because the item
  // endpoint samples a wider window than the hero one and there is no honest
  // match count to divide by.
  const items = d.items.filter((i) => i.share >= 12).slice(0, 24);

  const showingItems = META.dl.tab === "items";

  root.innerHTML = `
    <section class="home-section" style="padding-top:0">
      <div class="home-head">
        <h2 class="home-title">The meta right now</h2>
        <div class="home-meta">ranked matches · ${metaUpdatedAt(META.dl.loadedAt)}</div>
        <button class="link-btn" data-meta-retry type="button">${META.dl.loading ? "Refreshing…" : "Refresh"}</button>
      </div>
      ${railHtml([
        { label: "Sample", value: metaCount(d.matches), sub: "matches analysed" },
        { label: "Strongest", value: escapeHtml(top ? top.name : "—"), sub: top ? metaPct(top.winRate) + " win rate" : "" },
        {
          label: "Most picked",
          value: escapeHtml(contested ? contested.name : "—"),
          sub: contested ? `${metaPct(contested.pickRate, 0)} of matches` : "",
        },
      ])}
      <p class="hint" style="margin-top:12px;max-width:74ch">
        From the community Deadlock API, which is where every Deadlock number
        in this app comes from — Valve publishes no feed of its own.
      </p>
    </section>

    <section class="home-section">
      <div class="home-head">
        <h2 class="home-title">${showingItems ? "Items" : "Heroes"}</h2>
        <div class="home-meta">${showingItems ? `${items.length} shown` : `${heroes.length} heroes`}</div>
      </div>

      <div class="chip-row" style="margin-bottom:14px">
        <button class="chip ${!showingItems ? "selected" : ""}" data-dlmeta-tab="heroes" type="button">Heroes</button>
        <button class="chip ${showingItems ? "selected" : ""}" data-dlmeta-tab="items" type="button">Items</button>
        ${
          showingItems
            ? ""
            : `<span style="flex:1"></span>
               <button class="chip ${META.dl.sort === "winRate" ? "selected" : ""}" data-dlmeta-sort="winRate" type="button">Win rate</button>
               <button class="chip ${META.dl.sort === "pickRate" ? "selected" : ""}" data-dlmeta-sort="pickRate" type="button">Picked</button>`
        }
      </div>

      ${
        showingItems
          ? `<div class="meta-head items">
               <span>Item</span><span class="num">Win rate</span><span></span>
               <span class="num">Matches</span><span class="num">Bought</span>
             </div>
             ${items
               .map(
                 (i) => `
               <div class="meta-row items">
                 <div class="meta-name">${escapeHtml(i.name)}</div>
                 <span class="num meta-wr ${i.winRate >= 50 ? "good" : "bad"}">${metaPct(i.winRate)}</span>
                 ${metaBar(i.winRate)}
                 <span class="num">${metaCount(i.matches)}</span>
                 <span class="num">${i.buyMinute ? `${i.buyMinute.toFixed(0)} min` : "—"}</span>
               </div>`
               )
               .join("")}
             <p class="hint" style="margin-top:12px;max-width:74ch">
               <b>Bought</b> is the average minute the item goes down, which the
               Deadlock API publishes and the Dota side has no equivalent for.
               <b>Matches</b> is a raw count, not a percentage: the item
               endpoint samples a wider window than the hero one, so there is
               no honest figure to divide it by. Only the common items are
               listed &mdash; anything under an eighth as popular as the
               busiest item is left out.
             </p>`
          : `<div class="meta-head dl">
               <span></span><span>Hero</span><span class="num">Win rate</span><span></span>
               <span class="num">Picked</span><span class="num">KDA</span><span class="num">Souls</span>
             </div>
             ${heroes
               .map(
                 (h) => `
               <div class="meta-row dl" data-dlmeta-hero="${escapeHtml(h.name)}">
                 ${h.image ? `<img class="dl-hero-img sm" src="${escapeHtml(h.image)}" alt="" loading="lazy" decoding="async" />` : `<div class="dl-hero-img sm"></div>`}
                 <div class="meta-name">${escapeHtml(h.name)}</div>
                 <span class="num meta-wr ${h.winRate >= 50 ? "good" : "bad"}">${metaPct(h.winRate)}</span>
                 ${metaBar(h.winRate)}
                 <span class="num">${metaPct(h.pickRate, 0)}</span>
                 <span class="num">${h.kda.toFixed(2)}</span>
                 <span class="num">${dlFmtSouls(h.avgSouls)}</span>
               </div>`
               )
               .join("")}`
      }
    </section>`;

  root.querySelectorAll("[data-dlmeta-tab]").forEach((el) =>
    el.addEventListener("click", () => {
      META.dl.tab = el.dataset.dlmetaTab;
      renderDeadlockMeta();
    })
  );
  root.querySelectorAll("[data-dlmeta-sort]").forEach((el) =>
    el.addEventListener("click", () => {
      META.dl.sort = el.dataset.dlmetaSort;
      renderDeadlockMeta();
    })
  );
  root.querySelectorAll("[data-meta-retry]").forEach((el) =>
    el.addEventListener("click", () => loadDeadlockMeta(true))
  );
  root.querySelectorAll("[data-dlmeta-hero]").forEach((el) =>
    el.addEventListener("click", () => {
      DL.heroFilter = el.dataset.dlmetaHero;
      setView("dlmatches");
    })
  );
}

File: TheTracker/src-tauri/src/meta.rs
//! Meta: which heroes are strong right now, and which way they are moving.
//!
//! # Inspiration, and the line
//!
//! The shape of these pages is borrowed from dota2protracker and Statlocker
//! — strongest heroes grouped by role, a rating you can sort by, items that
//! are trending, a sense of the patch. That framing is worth learning from.
//!
//! None of the *data* comes from either site. Their numbers are collected
//! and rated by them, and that collection is the product; D2PT's rating in
//! particular is their own model over their own high-MMR sample. Taking it
//! would be lifting the thing they built. So everything below is computed
//! here, from the same public APIs the rest of the app already reads:
//!
//! - **Dota** — OpenDota's `/heroStats`, which reports picks and wins per
//!   hero across public matches, per rank bracket, for pro games, and as a
//!   seven-bucket trend.
//! - **Deadlock** — deadlock-api's `/v1/analytics/hero-stats` and
//!   `/v1/analytics/item-stats`.
//!
//! Where the public data cannot answer something the reference sites show,
//! the app says so rather than inventing it. OpenDota publishes Dota's role
//! *tags* (Carry, Support, Nuker…), not positions one through five, so the
//! grouping here is by tag and is labelled that way.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use serde::Serialize;

/// Meta moves on the scale of days, so a cached snapshot is fine and keeps
/// the app off the network on every tab switch.
const TTL: Duration = Duration::from_secs(60 * 30);

#[derive(Default)]
pub struct MetaCache {
    dota: Option<(Instant, DotaMeta)>,
    deadlock: Option<(Instant, DeadlockMeta)>,
}

pub type SharedMeta = Arc<Mutex<MetaCache>>;

// ---------- Dota ----------

#[derive(Debug, Clone, Serialize)]
pub struct MetaHero {
    pub id: u32,
    pub name: String,
    /// Slug for Valve's portrait CDN, e.g. "queenofpain".
    pub slug: String,
    pub roles: Vec<String>,
    /// Public matches the hero was picked in, over OpenDota's sample.
    pub picks: u64,
    #[serde(rename = "winRate")]
    pub win_rate: f64,
    /// Share of matches the hero appears in. Ten heroes are picked per
    /// match, so this is picks over (total picks / 10) — the number people
    /// mean by "contest rate", not a share of all picks.
    #[serde(rename = "pickRate")]
    pub pick_rate: f64,
    /// Win-rate change across the sample window, in points. Positive means
    /// the hero has been winning more lately.
    pub trend: f64,
    /// Immortal/Divine bracket win rate, where OpenDota has a sample.
    #[serde(rename = "highWinRate")]
    pub high_win_rate: Option<f64>,
    #[serde(rename = "proPicks")]
    pub pro_picks: u64,
    #[serde(rename = "proBans")]
    pub pro_bans: u64,
    #[serde(rename = "proWinRate")]
    pub pro_win_rate: Option<f64>,
    #[serde(rename = "turboWinRate")]
    pub turbo_win_rate: Option<f64>,
}

#[derive(Debug, Clone, Serialize)]
pub struct DotaMeta {
    /// Public matches behind the sample, derived from total picks.
    pub matches: u64,
    pub heroes: Vec<MetaHero>,
    /// Role tags present, in the order the UI should offer them.
    pub roles: Vec<String>,
}

/// Win rate over a slice of the trend buckets, as a percentage.
fn bucket_win_rate(wins: &[u64], picks: &[u64], from: usize, to: usize) -> Option<f64> {
    let w: u64 = wins.get(from..to)?.iter().sum();
    let p: u64 = picks.get(from..to)?.iter().sum();
    if p == 0 {
        return None;
    }
    Some(w as f64 / p as f64 * 100.0)
}

fn numbers(v: Option<&serde_json::Value>) -> Vec<u64> {
    v.and_then(|x| x.as_array())
        .map(|a| a.iter().filter_map(|n| n.as_u64()).collect())
        .unwrap_or_default()
}

pub async fn dota(cache: &SharedMeta) -> Result<DotaMeta, String> {
    {
        let c = cache.lock().unwrap();
        if let Some((at, ref meta)) = c.dota {
            if at.elapsed() < TTL {
                return Ok(meta.clone());
            }
        }
    }

    let value = crate::dota_api::get_json_public("/heroStats").await?;
    let rows = value.as_array().ok_or("OpenDota returned an unexpected shape")?;

    let total_picks: u64 = rows.iter().filter_map(|r| r.get("pub_pick").and_then(|v| v.as_u64())).sum();
    // Ten heroes are picked per match.
    let matches = (total_picks / 10).max(1);

    let mut heroes: Vec<MetaHero> = rows
        .iter()
        .filter_map(|r| {
            let picks = r.get("pub_pick").and_then(|v| v.as_u64())?;
            if picks == 0 {
                return None;
            }
            let wins = r.get("pub_win").and_then(|v| v.as_u64()).unwrap_or(0);
            let name = r.get("localized_name").and_then(|v| v.as_str())?.to_string();
            let raw = r.get("name").and_then(|v| v.as_str()).unwrap_or_default();
            let slug = raw.strip_prefix("npc_dota_hero_").unwrap_or(raw).to_string();

            let pick_trend = numbers(r.get("pub_pick_trend"));
            let win_trend = numbers(r.get("pub_win_trend"));
            // The final bucket is the week in progress and is always short,
            // so comparing it against the rest would read every hero as
            // falling. It is dropped and the remainder split in half.
            let usable = pick_trend.len().saturating_sub(1);
            let trend = if usable >= 2 && win_trend.len() >= usable {
                let mid = usable / 2;
                match (
                    bucket_win_rate(&win_trend, &pick_trend, 0, mid),
                    bucket_win_rate(&win_trend, &pick_trend, mid, usable),
                ) {
                    (Some(earlier), Some(recent)) => recent - earlier,
                    _ => 0.0,
                }
            } else {
                0.0
            };

            let bracket = |n: &str| -> Option<(u64, u64)> {
                let p = r.get(format!("{n}_pick")).and_then(|v| v.as_u64())?;
                let w = r.get(format!("{n}_win")).and_then(|v| v.as_u64())?;
                (p > 0).then_some((p, w))
            };
            // Bracket 8 is Immortal and is often empty in the snapshot; 7 is
            // Divine and is not. Prefer the higher one that actually has data.
            let high = bracket("8").or_else(|| bracket("7"));

            let pro_picks = r.get("pro_pick").and_then(|v| v.as_u64()).unwrap_or(0);
            let pro_wins = r.get("pro_win").and_then(|v| v.as_u64()).unwrap_or(0);
            let turbo_picks = r.get("turbo_picks").and_then(|v| v.as_u64()).unwrap_or(0);
            let turbo_wins = r.get("turbo_wins").and_then(|v| v.as_u64()).unwrap_or(0);

            Some(MetaHero {
                id: r.get("id").and_then(|v| v.as_u64()).unwrap_or(0) as u32,
                name,
                slug,
                roles: r
                    .get("roles")
                    .and_then(|v| v.as_array())
                    .map(|a| a.iter().filter_map(|s| s.as_str().map(String::from)).collect())
                    .unwrap_or_default(),
                picks,
                win_rate: wins as f64 / picks as f64 * 100.0,
                pick_rate: picks as f64 / matches as f64 * 100.0,
                trend,
                high_win_rate: high.map(|(p, w)| w as f64 / p as f64 * 100.0),
                pro_picks,
                pro_bans: r.get("pro_ban").and_then(|v| v.as_u64()).unwrap_or(0),
                pro_win_rate: (pro_picks > 0).then(|| pro_wins as f64 / pro_picks as f64 * 100.0),
                turbo_win_rate: (turbo_picks > 0).then(|| turbo_wins as f64 / turbo_picks as f64 * 100.0),
            })
        })
        .collect();

    heroes.sort_by(|a, b| b.win_rate.partial_cmp(&a.win_rate).unwrap_or(std::cmp::Ordering::Equal));

    // Role tags, most common first, so the filter row leads with the ones
    // that actually have heroes behind them.
    let mut counts: HashMap<String, usize> = HashMap::new();
    for h in &heroes {
        for r in &h.roles {
            *counts.entry(r.clone()).or_default() += 1;
        }
    }
    let mut roles: Vec<(String, usize)> = counts.into_iter().collect();
    roles.sort_by(|a, b| b.1.cmp(&a.1).then(a.0.cmp(&b.0)));

    let meta = DotaMeta {
        matches,
        heroes,
        roles: roles.into_iter().map(|(r, _)| r).collect(),
    };

    cache.lock().unwrap().dota = Some((Instant::now(), meta.clone()));
    Ok(meta)
}

// ---------- Deadlock ----------

#[derive(Debug, Clone, Serialize)]
pub struct MetaDlHero {
    pub id: u32,
    pub name: String,
    pub image: Option<String>,
    pub matches: u64,
    #[serde(rename = "winRate")]
    pub win_rate: f64,
    #[serde(rename = "pickRate")]
    pub pick_rate: f64,
    pub kda: f64,
    #[serde(rename = "avgSouls")]
    pub avg_souls: u64,
    #[serde(rename = "avgDamage")]
    pub avg_damage: u64,
}

#[derive(Debug, Clone, Serialize)]
pub struct MetaDlItem {
    pub id: u64,
    pub name: String,
    /// Matches the item was bought in. Deliberately a raw count, not a
    /// percentage: the item endpoint covers a different and much larger
    /// sample than the hero endpoint — its busiest item appears in roughly
    /// six times more matches than the hero sample contains — so dividing
    /// one by the other produced "pick rates" over 100%. Rather than invent
    /// a denominator, the count is shown as what it is.
    pub matches: u64,
    #[serde(rename = "winRate")]
    pub win_rate: f64,
    /// How common this item is next to the most-bought one, 0–100. A share
    /// of a known maximum is honest in a way a fabricated pick rate is not.
    pub share: f64,
    /// Average minute the item is bought. Deadlock's own API publishes this
    /// and nothing on the Dota side has an equivalent — when an item is
    /// bought says as much as how often.
    #[serde(rename = "buyMinute")]
    pub buy_minute: Option<f64>,
}

#[derive(Debug, Clone, Serialize)]
pub struct DeadlockMeta {
    pub matches: u64,
    pub heroes: Vec<MetaDlHero>,
    pub items: Vec<MetaDlItem>,
}

pub async fn deadlock(cache: &SharedMeta, hero_cache: &crate::deadlock::SharedHeroes) -> Result<DeadlockMeta, String> {
    {
        let c = cache.lock().unwrap();
        if let Some((at, ref meta)) = c.deadlock {
            if at.elapsed() < TTL {
                return Ok(meta.clone());
            }
        }
    }

    // Four requests, and one of them (the item asset list) is 726 entries.
    // Run sequentially the page sat on "Reading the meta…" for the better
    // part of half a minute, and the asset call timed out often enough that
    // items fell back to showing their numeric ids. Concurrently it is as
    // slow as the slowest one rather than the sum of all four.
    let (hero_rows, names, item_rows, item_names) = tokio::join!(
        crate::deadlock::get_json_public("/v1/analytics/hero-stats"),
        crate::deadlock::heroes(hero_cache),
        crate::deadlock::get_json_public("/v1/analytics/item-stats"),
        crate::popular::deadlock_item_names(),
    );
    let hero_rows = hero_rows?;

    let rows = hero_rows.as_array().ok_or("Deadlock API returned an unexpected shape")?;
    let total: u64 = rows.iter().filter_map(|r| r.get("matches").and_then(|v| v.as_u64())).sum();
    // Twelve players per Deadlock match.
    let matches = (total / 12).max(1);

    let num = |r: &serde_json::Value, k: &str| r.get(k).and_then(|v| v.as_u64()).unwrap_or(0);

    let mut heroes: Vec<MetaDlHero> = rows
        .iter()
        .filter_map(|r| {
            let played = num(r, "matches");
            if played == 0 {
                return None;
            }
            let id = num(r, "hero_id") as u32;
            let deaths = num(r, "total_deaths").max(1);
            Some(MetaDlHero {
                id,
                name: names.get(&id).map(|h| h.name.clone()).unwrap_or_else(|| format!("Hero {id}")),
                image: names.get(&id).and_then(|h| h.image.clone()),
                matches: played,
                win_rate: num(r, "wins") as f64 / played as f64 * 100.0,
                pick_rate: played as f64 / matches as f64 * 100.0,
                kda: (num(r, "total_kills") + num(r, "total_assists")) as f64 / deaths as f64,
                avg_souls: num(r, "total_net_worth") / played,
                avg_damage: num(r, "total_player_damage") / played,
            })
        })
        .collect();
    heroes.sort_by(|a, b| b.win_rate.partial_cmp(&a.win_rate).unwrap_or(std::cmp::Ordering::Equal));

    let items = deadlock_items(item_rows.unwrap_or(serde_json::Value::Null), &item_names);

    let meta = DeadlockMeta { matches, heroes, items };
    cache.lock().unwrap().deadlock = Some((Instant::now(), meta.clone()));
    Ok(meta)
}

fn deadlock_items(value: serde_json::Value, names: &HashMap<u64, String>) -> Vec<MetaDlItem> {
    let Some(rows) = value.as_array() else { return Vec::new() };

    let busiest = rows
        .iter()
        .filter_map(|r| r.get("matches").and_then(|v| v.as_u64()))
        .max()
        .unwrap_or(1)
        .max(1);

    let mut items: Vec<MetaDlItem> = rows
        .iter()
        .filter_map(|r| {
            let played = r.get("matches").and_then(|v| v.as_u64())?;
            if played == 0 {
                return None;
            }
            let id = r.get("item_id").and_then(|v| v.as_u64())?;
            let wins = r.get("wins").and_then(|v| v.as_u64()).unwrap_or(0);
            Some(MetaDlItem {
                id,
                // An unresolved name means the asset list did not load, not
                // that the item is nameless — say so rather than printing a
                // bare id and hoping nobody notices.
                name: names.get(&id).cloned().unwrap_or_else(|| "Name unavailable".to_string()),
                matches: played,
                win_rate: wins as f64 / played as f64 * 100.0,
                share: played as f64 / busiest as f64 * 100.0,
                buy_minute: r
                    .get("avg_buy_time_s")
                    .and_then(|v| v.as_f64())
                    .filter(|s| *s > 0.0)
                    .map(|s| s / 60.0),
            })
        })
        .collect();

    items.sort_by(|a, b| b.win_rate.partial_cmp(&a.win_rate).unwrap_or(std::cmp::Ordering::Equal));
    items
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_partial_final_bucket_is_excluded_from_the_trend() {
        // OpenDota's last trend bucket is the week in progress, so it is
        // always short. Including it would read as every hero falling at
        // once, which is the tell that the window is wrong rather than the
        // meta having moved.
        let picks = vec![1000, 1000, 1000, 1000, 1000, 1000, 90];
        let wins = vec![500, 500, 500, 600, 600, 600, 10];

        let usable = picks.len() - 1;
        let mid = usable / 2;
        let earlier = bucket_win_rate(&wins, &picks, 0, mid).unwrap();
        let recent = bucket_win_rate(&wins, &picks, mid, usable).unwrap();

        assert_eq!(earlier, 50.0);
        assert_eq!(recent, 60.0);
        assert_eq!(recent - earlier, 10.0, "a hero winning 10 points more lately should read as +10");
    }

    #[test]
    fn a_flat_hero_has_no_trend() {
        let picks = vec![500u64; 7];
        let wins = vec![250u64; 7];
        let usable = picks.len() - 1;
        let mid = usable / 2;
        let earlier = bucket_win_rate(&wins, &picks, 0, mid).unwrap();
        let recent = bucket_win_rate(&wins, &picks, mid, usable).unwrap();
        assert_eq!(recent - earlier, 0.0);
    }

    #[test]
    fn an_empty_window_yields_no_win_rate_rather_than_a_division_by_zero() {
        assert_eq!(bucket_win_rate(&[0, 0], &[0, 0], 0, 2), None);
        assert_eq!(bucket_win_rate(&[1], &[2], 5, 9), None);
    }
}
